File: packages/installer-server/src/lib/steps/persistApiKey.ts
import { setSecret } from '../secretsStore.js';
import type { InstallerStep } from './index.js';

export const persistApiKeyStep: InstallerStep = {
  id: 'persist-api-key',
  label: 'Store API key in OS keychain',
  shouldRun: ({ submission }) => {
    if (submission.provider === 'ollama' || submission.provider === 'lmstudio') return false;
    // Empty apiKey in settings-mode means "keep existing key" — skip the rotation.
    return submission.apiKey.trim().length > 0;
  },
  async run({ submission, bus }) {
    const account = `provider:${submission.provider}`;
    const { storage } = await setSecret(account, submission.apiKey);
    bus.log(persistApiKeyStep.id, 'info', `Saved API key (${storage}) as ${account}.`);
  },
};

File: internal/installer/steps/persist_api_key.go
// Package steps implements the individual pipeline operations the installer
// runs. Each step is a small struct so they can be composed differently
// (e.g. a future "settings reapply" mode that skips the heavy install/onboard
// steps) without duplicating logic.
package steps

import (
	"context"
	"fmt"
	"strings"

	"github.com/SolsticeSauer/solsclaw_beta/internal/installer"
	"github.com/SolsticeSauer/solsclaw_beta/internal/secrets"
)

type PersistAPIKey struct{}

func (PersistAPIKey) ID() string    { return "persist-api-key" }
func (PersistAPIKey) Label() string { return "Store API key in OS keychain" }

func (PersistAPIKey) ShouldRun(sc installer.StepContext) bool {
	if sc.Submission.Provider.IsLocal() {
		return false
	}
	// Empty key in settings-mode means "keep existing" — skip the rotation.
	return strings.TrimSpace(sc.Submission.APIKey) != ""
}

func (s PersistAPIKey) Run(_ context.Context, sc installer.StepContext) error {
	account := fmt.Sprintf("provider:%s", sc.Submission.Provider)
	_, storage, err := secrets.Set(account, sc.Submission.APIKey)
	if err != nil {
		return err
	}
	sc.Bus.Log(s.ID(), "info", fmt.Sprintf("Saved API key (%s) as %s.", storage, account))
	return nil
}
